### PR TITLE
Fix message collision check

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
There are some problems in the way that we check our messages when registering things into the descriptor pool. This PR addresses them by building and comparing `FileDescriptorProto` objects for the potential conflict. More details provided in summary of changes!

### Sample Repro Case
The serialized content for the same proto objects is not always consistent. Depending on how things are created & where they are validated things like `type_name` may not be fully qualified, and for some (mysterious) reason, there are false collision errors when nested types are used.

A simple repro case for the latter is outlined below:

Happy - no nested messages
```python3
from jtd_to_proto import jtd_to_proto

msg_name = "Foo"
package = "foo.bar"
schema = {"properties": {"foo": {"type": "boolean"}}}
jtd_to_proto(msg_name,package, schema)
# Registration with the descriptor pool is idempotent, which is expected
jtd_to_proto(msg_name,package, schema)
```

Not happy; raises `ValueError: Cannot add new file foo.proto to descriptor pool, file already exists with different content`
```python3
from jtd_to_proto import jtd_to_proto

msg_name = "Foo"
package = "foo.bar"
schema = {"properties": {"foo": {"properties": {"bar": {"type": "boolean"}}}}}
jtd_to_proto(msg_name,package, schema)
# Breaks unexpectedly
jtd_to_proto(msg_name,package, schema)
```
Note that this case is handled in `test_jtd_to_proto_duplicate_message`. 

### Summary of Changes
Two FileDescriptors are considered equal if their protos satisfy the following:
- They have the same package
- They have the same dependency
- All of their enums are well-aligned, i.e., they have the same `enum_type` names, and for any/all of those, every enum val/number lines up correctly
- All of their messages are well-aligned, i.e., they have the same `message_type` names, and for each message, every field has the same `label`, `type`, and `type_name`. For the last one, we provide some flexibility in case we're comparing something that isn't fully qualified to something that is

### CODE REVIEW

File: jtd_to_proto/jtd_to_proto.py [hunk 1]
```diff
@@ -30,6 +30,188 @@ def _to_upper_camel(snake_str: str) -> str:
...
+    """
+    if d1_type_name == d2_type_name:
+        return True
+    if len(d2_type_name) > len(d1_type_name):
+        d2_type_name, d1_type_name = d1_type_name, d2_type_name
+    return (
+        d1_type_name.startswith(".")
+        and d1_type_name.endswith(d2_type_name)
+        and not d2_type_name.startswith(".")
+    )
```
Comment (joerunde) [inline jtd_to_proto/jtd_to_proto.py:212]: would this give a false match for e.g.
```
_have_same_type_name(".foo.bar.baz.Buzz", "bar.baz.Buzz")
```
Since the first name starts with ".", ends with the second name, and the second does not start with "."?

We could `lstrip(".")` off both and check equality? Or assert that `abs(len(d1_type_name) - len(d2_type_name)) <= 1`?

File: tests/test_jtd_to_proto.py [hunk 2]
```diff
@@ -837,9 +891,111 @@ def test_jtd_to_proto_duplicate_message_name(temp_dpool):
...
 
 def test_to_upper_camel_empty():
     """Make sure _to_upper_camel is safe with an empty string"""
     assert _to_upper_camel("") == ""
+
+
+def test_to_type_name_validation():
+    """Make sure _to_upper_camel is safe with an empty string"""
+    assert _have_same_type_name("Foo", "Foo")
+    assert _have_same_type_name(".foo.bar.Foo", "Foo")
```
Comment (joerunde) [inline tests/test_jtd_to_proto.py:1000]: 🤔 related to the other comment, should we say that these are the same type? WOrth talkin in poerson about to get your context

File: jtd_to_proto/jtd_to_proto.py (hunk 1 above)
Comment (alex-jw-brooks) [reply jtd_to_proto/jtd_to_proto.py:212]: Fair point - it's not just starting `Foo` vs `.Foo`. I _think_ the fully qualified path ends up as

`<package name>.<Message name>.<Field name>`

Once it's been normalized. For example, if you have something like this:

```
from jtd_to_proto import jtd_to_proto

msg_name = "Message"
package = "package"
schema = {
    "properties": {
        "field": {"properties": {"another_field": {"type": "boolean"}}},
    }
}

descriptor = jtd_to_proto(
    msg_name,
    package,
    schema,
    validate_jtd=True,
)

descriptor = jtd_to_proto(
    msg_name,
    package,
    schema,
    validate_jtd=True,
)
```

and have it drop the things it's comparing here out, it's the raw field name against the package + message + field name. I.e,. `Field` vs. `.package.Message.Field`.

We definitely could tighten this to take the package name & message name as an arg so that the fully qualified path is more predictable, but probably would want an extra test to make sure this is safe when the field is a message declared in another package. What do you think?

File: jtd_to_proto/jtd_to_proto.py [hunk 3]
```diff
@@ -30,6 +30,152 @@ def _to_upper_camel(snake_str: str) -> str:
...
+        d1_field_descriptor = d1_field_descs[field_name]
+        d2_field_descriptor = d2_field_descs[field_name]
+        if (
+            d1_field_descriptor.label != d2_field_descriptor.label
+            or d1_field_descriptor.type != d2_field_descriptor.type
+        ):
+            return False
+    # For nested fields, we treat them similarly to how we've treated messages
+    # and recurse into comparisons used for the top level messages.
+    if d1.nested_type or d2.nested_type:
```
Comment (gabe-l-hart) [inline jtd_to_proto/jtd_to_proto.py:173]: We also need to check nested enums. I think that should be pretty straightforward to reuse the enum comparison above.

**Review by gabe-l-hart — CHANGES_REQUESTED** [review]
This is a much nicer solution (though unfortunate that it requires so much code!). I've got a couple of nits, but the main thing to fix is support for nested enums inside of messages. Otherwise it looks great

File: jtd_to_proto/jtd_to_proto.py [hunk 4]
```diff
@@ -30,6 +30,152 @@ def _to_upper_camel(snake_str: str) -> str:
...
+    """
+    d1_enum_descs = {enum.name: enum for enum in d1.enum_type}
+    d2_enum_descs = {enum.name: enum for enum in d2.enum_type}
+    if d1_enum_descs.keys() != d2_enum_descs.keys():
+        return False
+
+    for enum_name in d1_enum_descs.keys():
+        d1_enum_descriptor = d1_enum_descs[enum_name]
+        d2_enum_descriptor = d2_enum_descs[enum_name]
+        assert len(d1_enum_descriptor.value) == len(d2_enum_descriptor.value)
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply jtd_to_proto/jtd_to_proto.py:87]: Ah, nice catch - this is not supposed to be an assert; it should just be a `return False` since different lengths -> different enums -> different file descriptors

File: jtd_to_proto/jtd_to_proto.py (hunk 3 above)
Comment (alex-jw-brooks) [reply jtd_to_proto/jtd_to_proto.py:173]: Yup! Good catch, since `.enum_type` is just holding an iterable of EnumDescriptorProtos for both message descriptor protos & file descriptor protos, we can do it the same way. Made the change & added a test which previously was a failing case

File: jtd_to_proto/jtd_to_proto.py (hunk 1 above)
Comment (alex-jw-brooks) [reply jtd_to_proto/jtd_to_proto.py]: Marking this as resolved, since we discussed this in person and decided to drop the extra `type_name` stuff, since it's more restrictive than anything and may lead to some false positives

File: jtd_to_proto/jtd_to_proto.py [hunk 5]
```diff
@@ -30,6 +30,152 @@ def _to_upper_camel(snake_str: str) -> str:
...
+    return True
+
+
+def _check_message_descs_alignment(
+    d1_msg_container: Any, d2_msg_container: Any
+) -> bool:
+    """Determine if two message descriptor proto containers, i.e., RepeatedCompositeContainers
+    have the same message types. This means the following:
+
+    1. the messages contained in each FileDescriptorProto are the same.
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply jtd_to_proto/jtd_to_proto.py]: Done!

File: jtd_to_proto/jtd_to_proto.py [hunk 6]
```diff
@@ -30,6 +30,152 @@ def _to_upper_camel(snake_str: str) -> str:
...
+        bool
+            True if the contained message descriptor protos are identical.
+    """
+    d1_msg_descs = {msg.name: msg for msg in d1_msg_container}
+    d2_msg_descs = {msg.name: msg for msg in d2_msg_container}
+
+    # Ensure that our descriptors have the same dependencies & top level message types
+    if not d1_msg_descs.keys() == d2_msg_descs.keys():
+        return False
+    # For every encapsulated message descriptor, ensure that ever field has the same
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply jtd_to_proto/jtd_to_proto.py]: Done!

File: tests/test_jtd_to_proto.py [hunk 7]
```diff
@@ -837,6 +887,101 @@ def test_jtd_to_proto_duplicate_message_name(temp_dpool):
         )
 
 
+def test_jtd_to_proto_duplicate_enum_name(temp_dpool):
+    """Check that we cannot register a different message with the same name with wrong enum vals"""
+    msg_name = "Foo"
+    package = "foo.bar"
+    # The respective values we are going to register
+    first_enum_values = ["Hello", "World"]
+    second_enum_values = ["Hi", "Planet"]
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply tests/test_jtd_to_proto.py:910]: Good call - added another test for this

File: tests/test_jtd_to_proto.py [hunk 8]
```diff
@@ -837,6 +887,101 @@ def test_jtd_to_proto_duplicate_message_name(temp_dpool):
...
+            {
+                "properties": {
+                    "foo": {"properties": {"a_sad_nest": {"type": "int32"}}},
+                }
+            },
+            descriptor_pool=temp_dpool,
+            validate_jtd=True,
+        )
+
+
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply tests/test_jtd_to_proto.py:1058]: There are some gorpy edge cases that can come out of this that we talked about in person, mostly centered around file names / not being able to look things up easily in the descriptor pool based if something is compiled with conflicting defs via protoc in a file that's different from what JTD to Proto expects internally based on message name.

For now, I added a happy path test and several other sad path tests for some of grosser corners to ensure that things fail in the right ways in a way that matches the version of `protobuf` we are using. To avoid adding protoc to the dev environment, I just added the compiled bytes + registration to the descriptor pool calls to the tests directly, but let me know if you'd prefer to add protoc and compiled the .proto files as part of the dev environment instead

[the remark this answers is not in the document]
Comment (gabe-l-hart) [reply tests/test_jtd_to_proto.py:1058]: Sounds good!

**Review by gabe-l-hart — APPROVED** [review]
Ship it! Thanks Alex
